### ISSUE TITLE
Connect frontend to Books API (GraphQL)

### ISSUE THREAD

**Issue by knjk04** [Project-Books/book-project#772]
**Describe the solution you'd like**

![image](https://user-images.githubusercontent.com/11173328/135128777-365588a9-e9d5-4305-a035-58f89c5c6dce.png)

Design a new search page that looks like the above image. We need 'Search' in the navigation bar with the material icon. The search page should be responsive.

After a user has typed in a query and presses enter, call [findByTitle](https://github.com/Project-Books/books-api/blob/main/src/main/java/com/karankumar/booksapi/datafetchers/BookDataFetcher.java#L56), passing in the search query (we'll assume it's a title for now and then build upon this later). For now, we can console.log the response until we have finished our mockup designs for this page.

**Additional context**
- Be sure to read the contributing guidelines first!
- Discuss your implementation approach with us before we assign you to it
- Branch from `main` and set `main` as the destination branch in your pull request

**Comment by RivaD2** [comment]
@knjk04 Per slack convo, I am happy to take this on :)

**Comment by RivaD2** [comment]
@knjk04 My approach so far is: 
  - Add in another <NavItem /> to Nav.tsx 
  - Add in material search icon and add style to display them side by side
  - Create <Search /> component (thinking functional component using hooks if that works)
  - Add /search to Routes.tsx
  - Add <Search /> as new Route into App.tsx
  - Regarding query, I will make a separate method to handle this (not sure where it should live yet) and I am still thinking over best approach for this

**Comment by knjk04** [comment]
@RivaD2 This all sounds good!

**Comment by RivaD2** [comment]
@knjk04 Can you verify endpoint for this? 👍

**Comment by knjk04** [comment]
@RivaD2 There's only one endpoint in GraphQL, but see here: https://github.com/Project-Books/books-api/blob/main/src/main/java/com/karankumar/booksapi/datafetchers/BookDataFetcher.java#L56. You can call findByTitle and pass in the title of the book passed into the search bar

**Comment by RivaD2** [comment]
@knjk04 
   I was reading more about GraphQL (fun stuff but a lot happening here :) ) and I will for sure need an endpoint to hit so I can make the query.

   From the above images you sent, those are the query for the database, but I will need the full route that the frontend calls and also a sample of what the request body in JSON format would be. I think then I could possibly try and put something together.

I am doing more reading on this as this is my first time with GraphQL 👍

**Comment by RivaD2** [comment]
@knjk04 thank you, I will give this a try 👍

**Comment by RivaD2** [comment]
<img width="1185" alt="Screen Shot 2021-10-30 at 10 26 41 AM" src="https://user-images.githubusercontent.com/67027945/139543075-c92be191-2762-4698-bcaf-aad159b1cc4a.png">
<img width="739" alt="Screen Shot 2021-10-30 at 10 26 50 AM" src="https://user-images.githubusercontent.com/67027945/139543079-37595b2c-57d9-49ba-b60e-6eaf42c24a4b.png">


@knjk04 I was reviewing your screenshots in slack and have some questions. You are using built in GraphQL on postman but I am unable to see what the GET request looks like. 

I believe I have the query in React correct (I am using the same query from your screenshot on Slack) am getting a 404 because I am using POST.

Can you show an example of what the GET query looks like? I am guessing the payload goes as query params?

**Comment by knjk04** [comment]
@RivaD2 So sorry for the delay on this, things have been extremely hectic lately! 

![image](https://user-images.githubusercontent.com/11173328/142753779-d8d50134-2f82-40f5-8164-b8842514dd2e.png)

A couple of things I noticed when I ran this:

1) We're getting a 404 on port 8080. The Book Project runs on port 8080 and so does the Books API by default. I'll change the Books API to port 8082 by default. Were you running both the Book Project backend and the Books API? I think we both were getting 404s because the `graphql` endpoint isn't on the Book Project on port 8080.

2) I think this will still work even if we pass the bearer token into the Books API (it works for me on Postman), but we don't need to (we are looking to set up a different mechanism for authentication -- API key). Could you please change this?

Let me know how you get on!

**Comment by RivaD2** [comment]
@knjk04 I totally understand, no worries! Thank you for your help! 

I will take a closer look at this issue again this week and your thoughts on why we are getting the 404 make perfect sense to me :)  I will follow up you soon! 👍

**Comment by RivaD2** [comment]
@knjk04 I attempted to work on this but the default port is still 8080. Let me know know when you have updated the default port and I will pull in the changes. 👍

**Comment by knjk04** [comment]
https://github.com/Project-Books/books-api/blob/main/src/main/resources/application-dev.properties#L15

@RivaD2 Hmm, do you see the line above in the application-dev.properties file? Also, when the backend runs, do you see that in the log that it is using the dev environment?

**Comment by RivaD2** [comment]
@knjk04 Ohhh I see now, this is a separate repo. My apologies, I am workin
on it now. I ran the docker compose file and am now running `
BooksApiApplication.java` but I am getting `command not found`. Hmm, I
might be missing a step...

**Comment by knjk04** [comment]
> I have run the docker compose file and am now running BooksApiApplication.java but I am getting `command not found' (looking at the docs now on repo).

@RivaD2 What command did you run?

No, that's not a typo. Graphiql is the interactive playground we use for testing out queries/mutations

**Comment by RivaD2** [comment]
> > I have run the docker compose file and am now running BooksApiApplication.java but I am getting `command not found' (looking at the docs now on repo).
> 
> @RivaD2 What command did you run?
> 
> No, that's not a typo. Graphiql is the interactive playground we use for testing out queries/mutations

@knjk04 I followed instructions on README.md for running the app. I am trying to run `Run BooksApiApplication.java`

**Comment by RivaD2** [comment]
@knjk04 I am still unable to run the file mentioned above/build the project. Can you offer any guidance here?

I followed the instructions on the README.md but there aren't any instructions for how to build the project. I have tried different options but still no success.

It would be great if we could set up deployment of backend so I can focus on frontend tasks etc (I have lots of backend server questions). 👍

**Comment by knjk04** [comment]
@RivaD2 Sorry for the delay in getting back to you!

I'm not quite sure just yet. However, we are currently working on wrapping the backend and database with Docker, so hopefully that'll help (and if it doesn't, it should hopefully be easier for me to reproduce because we'll be using the same OS, Java versions, etc.). I appreciate it might be quite annoying if you're not able to get the backend working, so hopefully this will resolve things.

I looked into setting up a server on Azure, but this will have to wait for the time being because they're currently charging me a lot of money for services we hardly use. Once we have set up a non-profit (first, we'll be looking to change the name and trademark it), we can look into getting Azure open-source credits. 

TL;DR: a lot of work to go until a backend server is ready, but Docker should make things easier!

**Comment by RivaD2** [comment]
@knjk04 Totally understand Karan:) Thank you for getting back and letting me know about updates. Yes the roadblock I am hitting right now is I can't seem to run the app properly. Docker desktop is up and running fine, but when I try to run the `Run BooksApiApplication.java` file, nothing is happening.

Do you have any suggestions on this? I am not seeing the DGS codegen errors. I am seeing error 'command not found' so I am missing something here.

**Comment by knjk04** [comment]
@RivaD2 We've merged the Docker PR in, so try again using Docker. If it still doesn't work, could you please send some logs so I can look to reproduce this? (You may need to add this in, uncommented, into the Books API docker-compose file if it doesn't work for you on the Apple M1 chip: https://github.com/Project-Books/book-project/blob/0.2.0/backend/docker-compose.yml#L6)

**Comment by RivaD2** [comment]
@knjk04 Docker is running and graphiql is working fine, but I am still get a 403 on the OPTIONS request for `/graphql`. I am not seeing any logs in Docker. 

I have pushed my changes so you can check out my branch and run it through the frontend because the graphiql and Postman tools don't make options requests (but they are required in web browsers etc.)

See below:

<img width="1384" alt="Screen Shot 2021-12-26 at 11 17 10 AM" src="https://user-images.githubusercontent.com/67027945/147418020-2ecc9009-f27d-445c-8c59-6a375af045d9.png">
<img width="1047" alt="Screen Shot 2021-12-26 at 11 17 30 AM" src="https://user-images.githubusercontent.com/67027945/147418024-b56a550b-db1d-4df7-96b1-5b0f1f9684de.png">

Can you enable CORS for this route? 👍

**Comment by knjk04** [comment]
@RivaD2 I've set up CORS and disabled CSRF (for the time being), so hopefully it works now

**Comment by RivaD2** [comment]
@knjk04 Awesome, thank you! I will revisit this today 👍

**Comment by RivaD2** [comment]
@knjk04 I did a pull to get my branch even with main. I then did the clean install for maven and attempted to run the jar file on the book-project backend. I now get the error of: `java -jar target/book-project-0.2.0.jar`. 

I checked:
  - my editor
  - java version
  - path to file ( I don't see the jar file in /target directory, but I might be looking in the wrong place)

Can you confirm the proper path to file? If you happen to have run into this before and know some other steps I can take to troubleshoot, let me know. I will keep pluggin' away 👍

**Comment by knjk04** [comment]
> ![image](https://user-images.githubusercontent.com/11173328/148638330-c32e3d32-470f-42c6-81a2-813e67109276.png)

Thanks for bringing this up. I usually run it through my IDE, not Maven, so I wasn't aware. I'll look into this

That's strange. This is what I see: 

![image](https://user-images.githubusercontent.com/11173328/148638333-3dafe70c-0b07-4e49-818a-b722250e4c00.png)

**Comment by RivaD2** [comment]
@knjk04 I figured we might get some errors like that since I won't want to map over the data just yet. I can comment that out for now. I have a note in component explaining that I will console.log data for now (as ticket asks vs mapping)

**Comment by knjk04** [comment]
@RivaD2 Sorry for the delay. Here it is:

![image](https://user-images.githubusercontent.com/11173328/150393857-c0db1d52-5938-4017-87c6-acbca5f7f345.png)


![image](https://user-images.githubusercontent.com/11173328/150393791-10d94d78-6b29-4263-bb0e-5793522c0bd0.png)

**Comment by RivaD2** [comment]
@knjk04 Thank you for sending over the screenshots. I will take a look soon! 👍
